sycl/source/device_selector.cpp

```diff
@@ -28,8 +29,51 @@ static bool isDeviceOfPreferredSyclBe(const device &Device) {
          backend::level0;
 }
 
-device device_selector::select_device() const {
-  vector_class<device> devices = device::get_devices();
+// return a device with the requested deviceType (and requested backend)
+// if no such device is found, heuristic is used to select a device.
+// 'deviceType' is the desired device type
+//   info::device_type::all means it relies on the heuristic to select a device
+// 'be' is a specific desired GPU backend choice when multiple backends are
+// found.
+device device_selector::select_device(info::device_type deviceType,
+                                      backend be) const {
+  // return if a requested deviceType is found
+  if (deviceType != info::device_type::all) {
+    if (deviceType == info::device_type::host) {
+      vector_class<device> devices;
+      devices.resize(1);
+      return devices[0];
+    }
+
+    const vector_class<detail::plugin> &plugins = RT::initialize();
+    for (unsigned int i = 0; i < plugins.size(); i++) {
+      pi_uint32 numPlatforms = 0;
+      plugins[i].call<detail::PiApiKind::piPlatformsGet>(0, nullptr,
+                                                         &numPlatforms);
+      if (numPlatforms) {
+        vector_class<RT::PiPlatform> piPlatforms(numPlatforms);
+        plugins[i].call<detail::PiApiKind::piPlatformsGet>(
+            numPlatforms, piPlatforms.data(), nullptr);
+        for (const auto &piPlatform : piPlatforms) {
+          platform pltf = detail::createSyclObjFromImpl<platform>(
+              std::make_shared<detail::platform_impl>(piPlatform, plugins[i]));
+          if (!pltf.is_host()) {
+            vector_class<device> devices = pltf.get_devices(deviceType);
+            for (uint32_t i = 0; i < devices.size(); i++) {
+              if (deviceType != info::device_type::gpu) {
+                return devices[i];
+              } else if (devices[i].is_gpu() && be == pltf.get_backend()) {
+                return devices[i];
+              }
+            }
+          }
+        }
+      }
+    }
```

Comment:
I believe stack corruption will happen if no device is found as you return nothing here.

Reply:
The control reach this end of whole code section 'if (device_type != info::device_type::all) when the designated plugins cannot find the specified device_type. 
At this moment, we continue to the legacy heuristic based device selection based on points.
The user will get a WARNING because it will end up returning a different device than what the user requested.